### ISSUE TITLE
☂️  Tailwind class sorting lint rule

### ISSUE THREAD

**Issue by DaniGuardiola** [biomejs/biome#1274]
Update: I'm working on this! **Follow the updates in this Twitter thread:** https://twitter.com/daniguardio_la/status/1739715412131238122

### PRs

- https://github.com/biomejs/biome/pull/1362

### Description

Port the [TailwindCSS Prettier plugin](https://github.com/tailwindlabs/prettier-plugin-tailwindcss) that sorts Tailwind CSS utility classes. In Biome, it is being implemented as a lint rule that formats classes in JavaScript files though auto-fixes ([ref](https://github.com/biomejs/biome/discussions/164#discussioncomment-7291685)).

### Feature support / tasks checklist

> Features with an asterisk are not likely to be implemented as part of this task. They are listed for exhaustiveness, and considered in architectural decisions so that future support is possible, but they are not high priority.
>
> Features with a question mark are ideas that I'm not sure about, or I don't know enough about to determine if they make sense or how they could be implemented.

- Visitor:
  - [x] Custom functions (`clsx`, etc).
  - [x] Custom functions: tagged template support
  - [x] JSX attributes (`class` and `className`)
  - [x] Custom JSX attributes
  - [x] Object properties (e.g. in objects passed to `clsx`)
- Sorting:
  - [x] Utility classes
  - [x] Utility classes: arbitrary values
  - [x] Static variants
  - [x] Dynamic variants
  - [x] Arbitrary variants
  - [x] Arbitrary CSS
  - [ ] Parasite utilities
  - [ ] Screens e.g. `md:`, `max-md:` (simple config, px-only)
  - [ ] Screens (advanced config)
  - [ ] Negative values
  - [ ] With prefix
  - [ ] With important prefix (`!class`)
  - [ ] With custom separator
- Preset/config generation:
  - [ ] Tailwind CSS v3 (WIP)
  - [ ] Tailwind CSS v4
  - [ ] UnoCSS*
- Presets:
  - [x] Tailwind CSS preset
  - [ ] UnoCSS presets?*
- Config:
    - [x] Custom functions
    - [x] Custom JSX attributes
    - [x] Class order
    - [ ] Exclude classes?*
    - [ ] Utilities/layers
    - [ ] Variants
    - [ ] Screens
    - [ ] Preset (Tailwind CSS or no preset)
    - [ ] Extend utilities
    - [ ] Extend variants
    - [ ] Extend screens? (replace might be the only sane option)
    - [ ] Prefix
    - [ ] Separator
- Other features:
  - [ ] Automatic migration/sync `tailwind.config.js` -> `biome.json`
  - [ ] Resiliency to UnoCSS variant groups (naive sorting could mangle these) [docs](https://unocss.dev/transformers/variant-group)?*
  - [ ] UnoCSS variant groups sorting?*
  - [ ] UnoCSS attributify mode sorting?*
- Other tasks
  - [ ] Lint rule documentation
  - [ ] Config documentation
  - [ ] Migration documentation
- Polish:
  - [ ] PartialOrd / improved check perf
  - [ ] Single sort pass / lexicographical sort as last compare sort?
  - [ ] Better name for the rule (`useSortedUtilityClasses`?)
  - [ ] Avoid string allocations whenever possible.

## Upvote & Fund

- We're using [Polar.sh](https://polar.sh/biomejs) so you can upvote and help fund this issue.
- We receive the funding once the issue is completed & confirmed by you.
- Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/biomejs/biome/issues/1274">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/biomejs/biome/issues/1274/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/biomejs/biome/issues/1274/pledge.svg">
</picture>
</a>

**Comment by ematipico** [comment]
Hey @DaniGuardiola , thank you for opening this.

If you have the knowledge, do you know how this sorting thing works? Does Biome need to read the tailwind configuration file?

**Comment by DaniGuardiola** [comment]
@ematipico yeah I'm like 99% sure the config file is read. JavaScript is involved. Not sure how it fits in the biome way of doing things.

Is running JavaScript a possibility, at least to load the config file and transfer it to the rust program (or however that'd work)? or is it considered too slow?

**Comment by ematipico** [comment]
What I am interested in is the **business logic** of the plugin and how the configuration is involved in the sorting to the CSS classes.

I don't know Tailwind that well (as well as other people), so if someone with more knowledge can shed some light, it would help us to implement the feature in a faster way.

**Comment by DaniGuardiola** [comment]
@ematipico I will do a bit of research and get back to you.

**Comment by DaniGuardiola** [comment]
Leaving these as notes for the future.

This is where it all ultimately happens: https://github.com/tailwindlabs/tailwindcss/blob/master/src/lib/setupContextUtils.js#L930-L969

This function defers to this other function: https://github.com/tailwindlabs/tailwindcss/blob/master/src/lib/generateRules.js#L885-L930

The way sorting works is by leveraging the exact same mechanism that tailwind uses to sort the outputs. For example, `p-4 pt-6` and `pt-6 p-4` result in the same css, something like this:

```css
.p-4 {
  padding: 1rem;
}

.pt-6 {
  padding-top: 1.5rem;
}
```

So the class order after sorting will match this: `p-4 pt-6`.

This is explained here: https://tailwindcss.com/blog/automatic-class-sorting-with-prettier#how-classes-are-sorted

Regarding configuration. First let me disambiguate to avoid confusion. There are two separate concepts here:

- The Tailwind sorting options. This are strictly related to the sorting plugin. For example, one can indicate a list of functions whose string arguments (in function calls) will be interpreted as class names and sorted (e.g. `clsx`, `cva`, etc).
- The Tailwind CSS config. These are the options **for Tailwind itself**, like the theme, the files to process, the plugins, prefix, etc.

Sorting options are trivial, and just act as input for the formatting tool. What I'm talking about here is the latter, the involvement of the tailwind config in this process.

First, about this config: it is a JavaScript file that needs to run and be "resolved" (e.g. plugins are functions that need to call stuff like `addUtility` and such to register stuff for Tailwind to use. So, the final config is created by "resolving", a.k.a. executing a bunch of stuff like that.

Its role in sorting spans multiple things:

- General config like prefix (e.g. if the configured prefix is `tw`, classes will look like `tw-m-4`), the `! important` strategy, etc. These affect how classes are interpreted.
- The available utilities and variants, and their order (I'm guessing any of these coming from custom plugins are ordered just as the plugins are ordered in the config config file).

The prettier plugin seems to "hook" into all of the existing logic, essentially reusing the exact same mechanism to order the output. This makes sense, because why wouldn't they?

But I would argue that such complexity is not necessary, and biome could go with a way simpler approach based on simple heuristics and basic parsing/regexp.

I've done custom "reimplementations" of tailwind-like parsing for two projects now:

- https://github.com/DaniGuardiola/classy-ink
- https://github.com/compi-ui/tw-merge (complete rewrote of tailwind-merge using plain regexp)

So I would propose an approach like this + extensive testing. Of course, only using actual tailwind can guarantee that there won't be any false positives, but I think it is a very good compromise.

Regarding tailwind config, plugins and such, I would suggest that those options are moved to the formatter options. I think this biome feature should be generalize as a "utility class sorting" thing, with a tailwind preset.

So, in line with that, it should be simple enough to just add the utilities and variants that a plugin might add through config.


Sorry if this got a bit long / chaotic, I'm just researching and brainstorming to get things rolling. I'd be happy to start playing with this in a PR.

**Comment by DaniGuardiola** [comment]
To expand on the pros/cons of my proposed approach:

Pros

- Simpler. Like tens of thousands of LOC simpler.
- Covers 99.8% of cases. Yes, I made this number up, but I'm strongly convinced of it. I will explain that missing 0.2% and why it largely doesn't matter.
- It is generalizable to utility classes in general, which should cover other projects like nativewind, UnoCSS, etc.

Cons

- Can't use tailwindcss config directly. The user would need to re-specify these settings for the biome feature. Honestly, I think options like prefix are largely unused so I don't think it's a big deal. Regarding plugin-provided utilities and variants, the simplest solution is to provide a way to define these in the config as well. All in all, not the worst thing imo.
- False positives. Tailwind can tell the difference between something like `hover:px-4` and `hover:px-8268`. The first one is valid, and the second is not. This is because 4 is in the theme scale, and 8268 is not. The proposed approach would count both as valid, and sort them both. This is the 0.2% I was talking about. Anyone using `hover:px-8268` as a manually defined class should consider changing careers anyway. It's a non-issue.

**Comment by estubmo** [comment]
Yet to begin, because of events in my personal life.

**Comment by DaniGuardiola** [comment]
Thanks for the update. I'm gonna give it a go then :)

**Comment by Reed-Anderson** [comment]
Hey @DaniGuardiola , thanks so much for taking this on, it will be a huge upgrade for me. I hope this is an appropriate place to ask, is there any interest in supporting this popular tailwind/prettier request that was never implemented?

https://github.com/tailwindlabs/tailwindcss/discussions/7763

**Comment by itsezc** [comment]
> Cons
> 
> * Can't use tailwindcss config directly. The user would need to re-specify these settings for the biome feature. Honestly, I think options like prefix are largely unused so I don't think it's a big deal. Regarding plugin-provided utilities and variants, the simplest solution is to provide a way to define these in the config as well. All in all, not the worst thing imo.
> * False positives. Tailwind can tell the difference between something like `hover:px-4` and `hover:px-8268`. The first one is valid, and the second is not. This is because 4 is in the theme scale, and 8268 is not. The proposed approach would count both as valid, and sort them both. This is the 0.2% I was talking about. Anyone using `hover:px-8268` as a manually defined class should consider changing careers anyway. It's a non-issue.

First issue, can cause some friction for DX, and repeating configs is not ideal especially when there may be several devs in a team / project with an evolving tailwind config, but you're right its not the worst thing either.

Second, is a non issue as you've said. But, many people use arbitrary values i.e. `hover:px-[8rem]` or `hover:px-extra` if they're using a custom setup theme within their tailwind config. As long as these scenarios are handled to how the existing prettier plugin does, I don't think that'll be a huge issue.

**Comment by DaniGuardiola** [comment]
@Reed-Anderson thanks for bringing this up. While I share your pain, and think some tooling could definitely help, I think that's out of scope for this task.

I did see some approaches I liked in that thread, like one that turned multi-line strings into a normal, single line string on build (iirc). I don't think this task is the place to try something like that though.

**Comment by DaniGuardiola** [comment]
@itsezc good call-out. While I'm not going for perfection, I do wanna consider all of these cases, and I have some ideas to solve this that I'll share soon here for public feedback.

I'm a heavy tailwind user myself so I definitely want something that works across all reasonable use-cases.

**Comment by DaniGuardiola** [comment]
Here's my current thinking. Read my earlier messages on this thread for context.

## Tailwind CSS class sorting -> class sorting

While the main goal of this task is to provide a good replacement for the Prettier Tailwind CSS plugin, generalizing just makes sense to me, because other tools use the concept of utility classes as well. It also just seems like something that could be generally helpful, the ability to sort classes according to a configuration.

## Approach

The main challenge to solve is how the order of the classes is decided. Tailwind relies on its internal logic that also decides how the output is ordered, but we can't do that here, so we need a different approach. My thinking so far has been along the lines of having a config like this:

```jsonc
// simplified example, obviously
{
  "class-patterns": [
      // some kind of pseudo-regex / templating way of specifying these
     "p{y|x}-${scale}", // "special" value types like scale, numeric, etc
     "p-${scale}",
     "m{y|x}-${scale}",
     "m-${scale}",
     "grow$", // match "grow" without a suffix first ("exact")
     "grow-${numeric}" // then match other with a suffix
  ]
}
```

"Special" value types could be extended/customized:

```jsonc
{
  "types": {
    // add values
    "scale": { "extend": ["extra"] },
    // replace values
    "screens": { "values": ["md", "lg"] }
  }
}
```

The sorting algorithm would use this config to decide the order. Some other configs can be specified as well:

```jsonc
{
  "prefix": "tw-",
  // etc...
}
```

## Variants

Variants (`hover:active:focus:<utility class>`) also need to be taken into consideration, and they are ordered as well iirc. The order could be specified similarly. The separator (colon by default) could also be configurable.

## Presets

Of course, having and maintaining all of this in a config JSON file is annoying, so Biome would ship with presets for Tailwind and any other libraries that are popular enough.

```jsonc
{
  "preset": "tailwind-css",
}
```

## Manual configuration

As I described in a previous comment, a drawback of this approach is that there's a bunch of manual configuration to maintain, which has to be kept in sync with the Tailwind config. To alleviate this, it'd be doable to create a package that automatically reads the Tailwind config and syncs the Biome config to it. Doesn't necessarily need to be part of Biome itself, but it can be a simple npm package. E.g.

```
$ npm i -D sync-tailwind-biome
$ sync-tailwind-biome
```

I think this should make that pain much more manageable.

---

These are my thoughts so far. On the development side of things, my focus is going to be creating a minimal POC that works for a default tailwind config. This is the first time using Rust so it's enough to keep me busy for a while :P

Feedback is super welcome though, I'd like to polish these ideas so that I can be confident when it's time to implement the advanced features.

**Comment by ematipico** [comment]
> Tailwind CSS class sorting -> class sorting

Unfortunately I can't venture an opinion because I don't know much of the logic we are about to implement. Surely, I a big fan of consistency, so it's definitely a good start.

> Approach

What are the JSON snippets that you shared? I miss some context even though I read the previous messages. 

> Variants

I don't understand the context of this, can you expand a bit please?

> Manual configuration

I would suggest a different approach. I like having a separated package that does the work, and I think we can also have it under our organization, maintained by some volunteers. 

The package should return a JSON file via `stdout`, and then we can use this package in our command `biome migrate` The command can read the result of this package via `stdout`, and use it to update the configuration file. Here's why:
- Biome has already a parser and deserialiser for `biome.json`, let's use it.
- By using our internal deserializer, we can catch errors from the package very early.
- The package stays slim, and doesn't require any parser/regex to update the `biome.json`.
- Other tools shouldn't mess with `biome.json`. Using Biome itself is more trustworthy.

**Comment by tcoopman** [comment]
> * Honestly, I think options like prefix are largely unused so I don't think it's a big deal.

This comment is not completely clear to me. Does this mean that you won't provide support for prefixes or that they'll have to be specified in the biome config?

**Comment by DaniGuardiola** [comment]
@tcoopman the latter. In any case, the plan is to provide an automated way to update the config, and I'd expect that to be the most popular way to use this, so it's even less of a big deal.

**Comment by XiNiHa** [comment]
I wonder if it'd be possible to support UnoCSS transformer features like [variant groups](https://unocss.dev/transformers/variant-group)...

**Comment by johnpyp** [comment]
It might also be useful to look at https://github.com/avencera/rustywind, as this is a pure rust implementation of tailwind sorting which also doesn't look at the tailwind.config.js file afaict

**Comment by DaniGuardiola** [comment]
@XiNiHa it's possible, but it's a bit out of scope for now. Maybe in a future iteration.

@johnpyp thanks for the suggestion. I'm already very familiar with how Tailwind CSS sorting works, and an initial version is about to merge with only a couple of details missing, so there's no value in researching other projects anymore at this stage. Fwiw, I did take a look around that project but ended up using Tailwind CSS itself as the source of truth, which I believe is the best way to ensure correctness.

**Comment by GustavoBonfimS** [comment]
Could this rule be a safe action? Today we can only organize classes using --apply-unsafe. If it were a safe rule it could be automatically organized with the VS Code extension when the file is saved.

#1714

**Comment by DaniGuardiola** [comment]
This rule is a work in progress and there's still a lot of work to do before we can consider it stable. I like your linked suggestion about opting into unsafe fixes, and I understand that you want to use the rule with a nicer workflow, but please be patient.

**Comment by ematipico** [comment]
@jsonMartin @GustavoBonfimS 

Thank you for the engagement, but these kinds of comments don't bring any value to the conversation, they are out of scope. We have a plan and want to stick to it. 

You could help us by reporting possible bugs and improvements, or by implementing the missing pieces.

**Comment by BarakChamo** [comment]
@DaniGuardiola very exciting for this feature, sorting TW classes automatically is a big part of a predictable UI collaboration flow for us and many teams using TW.

Wanted to chime in to the discussion and mention that it appears the TW team is considering arbitrary value support for different utility classes without a square bracket format in TW 4. (i.e. there won't be a predefined scale of values but both `p-4` and `p-1000` would just work).

This would be in line with your suggested implementation and offer potential future proofing for dynamic values, while still auto-sorting current classes like, for example, `p-865` in current tailwind, which is not part of the theme. Seems like an outlier issue though.

**Comment by hilja** [comment]
I’m sure this falls under the undone tasks, but I had a little Kafkaesque experience trying to get the sorting works for this abomination. Biome stuck the classes together, the space seems to break the sorting:

![Screenshot 2024-03-09 at 16 16 26](https://github.com/biomejs/biome/assets/2558021/8aa1ae17-a455-448c-81fa-92831c011c6b)

If I used a `+` then the template literal rule kicks in :) 

![Screenshot 2024-03-09 at 16 33 28](https://github.com/biomejs/biome/assets/2558021/465785db-8410-4042-937c-13369e692fca)

The space again:

![Screenshot 2024-03-09 at 16 20 30](https://github.com/biomejs/biome/assets/2558021/8de6033e-4cd9-42b0-a0e4-5c83a0c193d9)

This works though:

![Screenshot 2024-03-09 at 16 29 57](https://github.com/biomejs/biome/assets/2558021/c7adf88f-e9d0-4f25-9cd0-e9f97653ef02)

Anyway this is a bad way of handling classes because now Tailwind won't know that `color.bg.medium` should be included in the final css. In general the sorting is a wonderful!

**Comment by sourcec0de** [comment]
Looks like Tailwind 4.0 builds its new oxide engine on Rust.

Might be useful to this thread.
https://youtu.be/CLkxRnRQtDE?t=2149
https://tailwindcss.com/blog/tailwindcss-v4-alpha

**Comment by aarvinr** [comment]
Could we add TailwindCSS class support for other languages like Vue or Svelte, instead of just JSX? Currently it doesn't seem to work, but it seems like it should be an easy fix since it's just `class` instead of `className`.

**Comment by DaniGuardiola** [comment]
@sourcec0de thanks for the suggestion. The rust part of tailwind v4 deals primarily with 1) parsing and 2) generation of CSS + the new CSS based config.

1 is redundant since Biome has its own parser which we use, and the tailwind specific lexer is already in place and works well.

2 is not necessary for the most part since we don't have any need to generate CSS. For the CSS based config, it might be useful. I'm waiting for things to stabilize in that area before doing anything about it, and I'm already in touch with the tailwind team so rest assured that if there's a chance to share any logic, it'll happen.

Most of the logic that is relevant for sorting is still JavaScript, so we need to reimplements and adapt it into Rust, as we've already done for some of it.

**Comment by DaniGuardiola** [comment]
@aarvinr class is already supported, and you can add any extra attributes/props through the options.

What is not supported is the Vue and Svelte languages as a whole. Once those are supported, we will make sure that this rule works with them.

**Comment by MH4GF** [comment]
Thanks for working on this issue 😄 Please allow me to report the bug.
Starting with v1.6.0, it appears that whitespace with placeholders in template literals will be removed. 

```
./apps/web/app/_components/Time.tsx:7:35 lint/nursery/useSortedClasses  FIXABLE  ━━━━━━━━━━━━━━━━━━━

  × These CSS classes should be sorted.
  
    5 │ export const Time = ({ children, className, ...props }: PropsWithChildren<Props>) => {
    6 │   return (
  > 7 │     <time className={`${className} text-sm text-zinc-500`} {...props}>
      │                                   ^^^^^^^^^^^^^^^^^^^^^^
    8 │       {children}
    9 │     </time>
  
  i Unsafe fix: Sort the classes.
  
    7 │ ····<time·className={`${className}·text-sm·text-zinc-500`}·{...props}>
      │                                   -                                   
```

If this space is removed, the correct class name cannot be retained.
This problem did not occur in `1.5.3-nightly.24fcf19`.

Attached is the CI job log for my project. I hope it will be helpful!
ref: https://github.com/MH4GF/mysite/actions/runs/8248800629/job/22559888722?pr=201

**Comment by reowl666** [comment]
In my environment, if I apply this rule and try to combine class names with a half-width space using join(), the half-width space will be deleted, so I would like to report this just in case.

<img width="844" alt="スクリーンショット 2024-03-31 12 40 35" src="https://github.com/biomejs/biome/assets/68766032/e845fcc5-eb58-4594-b91a-6af19a16d234">

**Comment by DaniGuardiola** [comment]
@reowl666 thank you for the report. Seems like we should just ignore any strings that don't contain any text, or that don't include at least one tailwind class.

**Comment by hangaoke1** [comment]
Could it detect the jsxQuoteStyle configuration ? 
```js
"formatter": {
	"quoteStyle": "single",
	"jsxQuoteStyle": "double",
}
```
![image](https://github.com/biomejs/biome/assets/19623160/fd3fd66a-0819-46ae-993e-6196a594639e)

**Comment by ematipico** [comment]
I removed @DaniGuardiola from the assignees because he doesn't have time anymore. If anyone wants to step up and help take this to the finish line, that would be great. 

You can help us with the logic, and we can help with the Rust code.

**Comment by DaniGuardiola** [comment]
Happy to guide wherever possible. Also, no promises, but I'm hoping I'll be able to contribute to this further at some point.

**Comment by Hidde-Heijnen** [comment]
> I removed @DaniGuardiola from the assignees because he doesn't have time anymore. If anyone wants to step up and help take this to the finish line, that would be great.
> 
> You can help us with the logic, and we can help with the Rust code.

Can you provide some more detail (or where people could find), what part of the logic would still be needed?

**Comment by lutaok** [comment]
Hi, I would like to take this on!

**Comment by Mardoxx** [comment]
Template literals are still broken [same as this](https://github.com/biomejs/biome/issues/1274#issuecomment-2001479201)

e.g. ```className={`relative flex-grow ${isVisible? "visible" : "hidden"}`}```

will fail and correct to ```className={`relative flex-grow${isVisible? "visible" : "hidden"}`}``` which is wrong.

Can't use twin.macro because remix

**Comment by ematipico** [comment]
@Mardoxx the fix will be released in the next release. If you can't wait, you can try the latest nightly. If the bug persists, please open a new issue, as comments here should be about developments and tend to be less visible unless someone is subscribed.

**Comment by afonsojramos** [comment]
Would be probably good to also add to biome all of the rules that https://github.com/francoismassart/eslint-plugin-tailwindcss supports, such as:
[enforces-negative-arbitrary-values](https://github.com/francoismassart/eslint-plugin-tailwindcss/blob/master/docs/rules/enforces-negative-arbitrary-values.md): make sure to use negative arbitrary values classname without the negative classname e.g. -top-[5px] should become top-[-5px]
[enforces-shorthand](https://github.com/francoismassart/eslint-plugin-tailwindcss/blob/master/docs/rules/enforces-shorthand.md): merge multiple classnames into shorthand if possible e.g. mx-5 my-5 should become m-5
[no-arbitrary-value](https://github.com/francoismassart/eslint-plugin-tailwindcss/blob/master/docs/rules/no-arbitrary-value.md): forbid using arbitrary values in classnames (turned off by default)
[no-custom-classname](https://github.com/francoismassart/eslint-plugin-tailwindcss/blob/master/docs/rules/no-custom-classname.md): only allow classnames from Tailwind CSS and the values from the whitelist option
[no-contradicting-classname](https://github.com/francoismassart/eslint-plugin-tailwindcss/blob/master/docs/rules/no-contradicting-classname.md): e.g. avoid p-2 p-3, different Tailwind CSS classnames (pt-2 & pt-3) but targeting the same property several times for the same variant.
[no-unnecessary-arbitrary-value](https://github.com/francoismassart/eslint-plugin-tailwindcss/blob/master/docs/rules/no-unnecessary-arbitrary-value.md): e.g. replacing m-[1.25rem] by its configuration based classname m-5

Which I would say are rather important!

**Comment by MrOxMasTer** [comment]
I am the only one with such a problem that he complains about the wrong order of classes (although in my opinion they were sorted by eslint before) and at the same time he does not format not through vs code, not through the cli

```
  ! These CSS classes should be sorted.

  > 21 │             className="w-full mt-6"
       │                       ^^^^^^^^^^^^^
    23 │           >

  i Unsafe fix: Sort the classes.

    21    │ - ············className="w-full·mt-6"
       21 │ + ············className='mt-6·w-full'
    23 23 │             >


Checked 1 file in 4ms. No fixes applied.
Found 1 warning.
```
![image](https://github.com/biomejs/biome/assets/59291123/c897956f-a780-4ee6-9e3b-712fea2ef6a5)

```
pnpm biome format src\...\component.tsx
Checked 1 file in 2ms. No fixes applied.
```

**Comment by Gomah** [comment]
> ```
> pnpm biome format src\...\component.tsx
> Checked 1 file in 2ms. No fixes applied.
> ```

You're running the formatter, not the linter, here's what you're looking for:
```
biome lint . --write --unsafe
```

Or if you want to run the formatter, linter and import sorting:

```
biome check . --write --unsafe
```

**Comment by DaniGuardiola** [comment]
Update: variant sorting (with some caveats) will be supported in the next version of Biome, thanks to the excellent work by @lutaok :)

**Comment by MrOxMasTer** [comment]
> ```
> biome check . --write --unsafe
> ```

unsafe really worked. How can I set unsafe in the extension itself for formatting?
![image](https://github.com/biomejs/biome/assets/59291123/21f42630-d06e-4e2a-ad72-ecf96044d129)

**Comment by DaniGuardiola** [comment]
@MrOxMasTer it's not possible. Related: https://github.com/biomejs/biome/issues/1274#issuecomment-1937834636

Please keep the discussion here on topic, for bug reports and specific discussion about the feature. Feel free to use Discord or GitHub Discussions for questions and discussion of more general topics.

**Comment by ematipico** [comment]
> > ```
> > biome check . --write --unsafe
> > ```
> 
> unsafe really worked. How can I set unsafe in the extension itself for formatting?
> ![image](https://github.com/biomejs/biome/assets/59291123/21f42630-d06e-4e2a-ad72-ecf96044d129)
> 
> 
> 

Please, search the website! Have everything in there: https://biomejs.dev/linter/#configure-the-rule-fix

@DaniGuardiola this is now possible since v1.8.0

**Comment by hangaoke1** [comment]
Whether it is possible to be compatible with Spaces
![image](https://github.com/biomejs/biome/assets/19623160/cb904d11-f297-462f-ab54-e800b1306709)

**Comment by MrOxMasTer** [comment]
guys, you haven't done the sorting yet, you can use the headwind extension if you work in vs code.

**Comment by MrOxMasTer** [comment]
> Please, search the website! We have everything in there: https://biomejs.dev/linter/#configure-the-rule-fix

If you are talking about the "fix" setting, which takes the values "safe" or "unsafe" for each rule, then I can say that I used it and it did not work for me

**Comment by ematipico** [comment]
@MrOxMasTer I advise you to open a new issue with a reproduction, this issue is for developments only and we don't want to ping subscribers with messages that are unrelated

**Comment by damianobarbati** [comment]
@DaniGuardiola other 2 things to consider on why the tailwind config is needed and possible problems:
1. A future needed to move away from eslint + tailwind plugin is the <https://github.com/francoismassart/eslint-plugin-tailwindcss/blob/master/docs/rules/no-custom-classname.md> rule which gives you error if the class you specified does not exist.
2. Some classes are automatically generated depending on the tailwind config file: for instance, if you declare a color `{"red":{"blood":"#F00"}}`, you can have use the class `fill-red-blood`.
3. Some classes are inside your `.css` files, and those are valid.
4. In a monorepo, you can have the tailwind config inside the single package folder and not in the root monorepo folder.

Hope this helps.

**Comment by Goldziher** [comment]
Running autofix with unsafe replaces all `className=` with `class=` inside react components.

**Comment by DaniGuardiola** [comment]
@Goldziher please provide a minimal reproduction.

**Comment by BorisZubchenko** [comment]
Seems like nobody has mentioned this before. Biome formats this:
```jsx
<section className="[grid-area:'body'] text-center">1</section>
```
To this:
```jsx
<section className='text-center [grid-area:'body']'>1</section>
```
Here, double quotes were replaced with single ones. This leads to a syntax error.

**Comment by ematipico** [comment]
@BorisZubchenko please open a bug report, this issue isn't for bug reports.

**Comment by vgseven** [comment]
add support for rules for un-supported or invalid class of tailwind, inspired by [eslint-plugin-tailwind](https://github.com/francoismassart/eslint-plugin-tailwindcss)

**Comment by warrenbhw** [comment]
Is there any way to apply the unsafe fix on save?

**Comment by ematipico** [comment]
@warrenbhw this question was already asked and answered in this very issue. Please search here or on the website: https://github.com/biomejs/biome/issues/1274#issuecomment-2203887336

**Comment by lobosan** [comment]
Hi, I'm migrating my Astro project to Biome 1.8.3, and I love the simple setup, the only thing remaining for me is Tailwind class sorting.

This is my biome.json
```
{
  "$schema": "./node_modules/@biomejs/biome/configuration_schema.json",
  "css": {
    "formatter": {
      "enabled": true
    },
    "linter": {
      "enabled": true
    },
    "parser": {
      "cssModules": true
    }
  },
  "files": {
    "ignore": [".astro", "dist", "node_modules"]
  },
  "formatter": {
    "formatWithErrors": true,
    "indentStyle": "space",
    "lineWidth": 100
  },
  "javascript": {
    "formatter": {
      "arrowParentheses": "asNeeded"
    }
  },
  "linter": {
    "rules": {
      "nursery": {
        "useSortedClasses": "error"
      }
    }
  }
}
```

This is my settings.json for VSCode
```
{
  "editor.codeActionsOnSave": {
    "quickfix.biome": "explicit",
    "source.organizeImports.biome": "explicit"
  },
  "editor.defaultFormatter": "biomejs.biome",
  "editor.formatOnSave": true
}
```

However, when I run
`npx @biomejs/biome check --unsafe `

I don't see any error in my .astro file that has the invalid example of the documentation
`<div class="px-2 foo p-4 bar" />`

Can someone please help me to fix the issue? Also, ideally I would love to sort the classes when saving.

**Comment by Trombach** [comment]
> Hi, I'm migrating my Astro project to Biome 1.8.3, and I love the simple setup, the only thing remaining for me is Tailwind class sorting.
> 
> This is my biome.json
> 
> ```
> {
>   "$schema": "./node_modules/@biomejs/biome/configuration_schema.json",
>   "css": {
>     "formatter": {
>       "enabled": true
>     },
>     "linter": {
>       "enabled": true
>     },
>     "parser": {
>       "cssModules": true
>     }
>   },
>   "files": {
>     "ignore": [".astro", "dist", "node_modules"]
>   },
>   "formatter": {
>     "formatWithErrors": true,
>     "indentStyle": "space",
>     "lineWidth": 100
>   },
>   "javascript": {
>     "formatter": {
>       "arrowParentheses": "asNeeded"
>     }
>   },
>   "linter": {
>     "rules": {
>       "nursery": {
>         "useSortedClasses": "error"
>       }
>     }
>   }
> }
> ```
> 
> This is my settings.json for VSCode
> 
> ```
> {
>   "editor.codeActionsOnSave": {
>     "quickfix.biome": "explicit",
>     "source.organizeImports.biome": "explicit"
>   },
>   "editor.defaultFormatter": "biomejs.biome",
>   "editor.formatOnSave": true
> }
> ```
> 
> However, when I run `npx @biomejs/biome check --unsafe `
> 
> I don't see any error in my .astro file that has the invalid example of the documentation `<div class="px-2 foo p-4 bar" />`
> 
> Can someone please help me to fix the issue? Also, ideally I would love to sort the classes when saving.

Biome can't yet parse the html template section of Astro files, so nothing gets formatted/linted in there, including class names. THis will only work once biome has full support for Astro files

**Comment by MarkLyck** [comment]
After upgrading to biome 1.9.0 the Quick Fix for sorting the classes stopped working for me.

Downgrading back to 1.8.2 worked, and I can quick fix my classes sorting again.

**Comment by dsbrianwebster** [comment]
> After upgrading to biome 1.9.0 the Quick Fix for sorting the classes stopped working for me.
> 
> Downgrading back to 1.8.2 worked, and I can quick fix my classes sorting again.

Same here

**Comment by ematipico** [comment]
@MarkLyck update to 1.9.1. That issue was reported already and fixed.

**Comment by columk1** [comment]
In 1.9.2 the `useSortedClasses` flag appears at the space between two template literals when they are used within a list of classes. The quick fix removes the space which causes the classes to be concatenated.

**Comment by ematipico** [comment]
@columk1 this is a development hub, not an issue tracker. The issue was already reported https://github.com/biomejs/biome/issues/4041 by another user who rightfully created an issue.

Please refrain from replying or adding comments that aren't related to the developments, because this will trigger the many users subscribed to the issue. Thank you! 🙏

**Comment by ematipico** [comment]
Hi all, 

Implementing the issue is going slower than we wished, but that's open source, right? We're struggling to find people who have in-depth knowledge of the domain (tailwind) and coding skills (Rust)—of course, and time, too. 

We want to try a different experiment this time. We are still determining if it will work, but at worst, we won't make any progress. 

If you have in-depth knowledge of at least one of the missing features you require (check the first comment of the issue), and you have time and will to help the project, we would ask you the following:
- Create an issue that is linked to this one. In the issue, you will be able to explain to us maintainers the feature to implement, as much technical as possible, as extensive as possible. Ideally, you'd want to define a potential algorithm for implementing the feature. Spare no details.
- Add **invalid** use cases paired with the expectations of the rule. The more use cases you provide, the better. We will use these use cases inside our test suite. Here's an example:
	```diff
	// use case 1
   - "class-1 class-3 class-4"
   + "class-4 class-3 class-1"
	// use case 2
   - "class-1 class-3 class-4"
   + "class-4 class-3 class-1"
	```
- One of the maintainers will ask you questions, and pick up the task. If you want, you can also be involved in the review of the task, but it's not mandatory. As long as you match the expectations of the use cases you provide, we should be good.


> [!NOTE]
> I increased the task pledge to 1k USD. This is a mastodontic task that requires a lot of work. Also, 70% of the bounty will be distributed to the contributors. Unfortunately, Polar seems buggy and can't update the issue badge, but you can check the [**link of the page**](https://polar.sh/biomejs/biome/issues/1274) to confirm.

> [!IMPORTANT]
> If you decide to help us, and you have extensive knowledge about a topic, you leave a comment here to notify possible other contributors who want to help with the same topic. We want to try to *avoid double issues*, so you can collaborate on the same issue.

**Comment by silvenon** [comment]
Regarding submitting test cases, will the initial goal be to pass prettier-plugin-tailwindcss's own test suite, or is the approach alternative to that?

**Comment by ematipico** [comment]
> Regarding submitting test cases, will the initial goal be to pass prettier-plugin-tailwindcss's own test suite, or is the approach alternative to that?
> 
> Btw, once I donated more to the issue Polar updated the increased task pledge as well in the card 🥳 

@silvenon Yes. It's my understanding that the correct ordering of classes is crucial in certain cases, so yes, we should their testing suite

**Comment by DaniGuardiola** [comment]
@ematipico I'd be happy to do this, fwiw. I had a call a while back with @lutaok where I basically did that, going very deep into the details of every item in the list. I think a format that would work is if I record a video while sharing my screen, so I can explain everything properly.

Typing would take me potentially weeks, but I can sit down and do a ~2 hour recording (what our call approximate took) going over everything. Would that help?

**Comment by ematipico** [comment]
@DaniGuardiola that's also another viable option, feel free to help. However, we still need to have valid and invalid cases for each feature that we must have, so we have some sort of harness.

**Comment by mariusch** [comment]
Hey! 👋 Just wanted to chime in with a quick observation about the `useSortedClasses` rule.

We're using Geist fonts in our Next.js app, and we noticed that this rule removes the space between font variable classes:

E.g
   ```typescript
   className={cn(`${GeistSans.variable} ${GeistMono.variable}`)}
   ```
   and 
   
  ```typescript:apps/rh/src/components/sak-status-badge.tsx
{/* biome-ignore lint/nursery/useSortedClasses: Removes space between classes and breaks them */}
<Badge variant="outline" className={`${config.style} ${badgeSize}`}>
```

That space is crucial for the classes to work correctly. Any chance we could update the rule to preserve spaces in cases like this? It'd save us from having to use ignore comments. 

Thanks for all your hard work on Biome! 🙌

**Comment by ematipico** [comment]
@mariusch 

Thank you for engaging, but this isn't an issue to track and report bugs. This is an issue to coordinate the work, and discuss implementation details. The issue you're facing was already _reported_ and **fixed**. Please, next time try to search for issues similar to yours, and _file a bug report_ if you can't find it.

**Comment by pvinis** [comment]
This rule is great, but it should also trim whitespace. So no spaces at the start of the line, no spaces at the end, and only one space between each classname.

**Comment by silvenon** [comment]
I'm guessing fixtures for handling whitespace are already included in `prettier-plugin-tailwindcss`' s test suite.

Note that this issue is for coordinating work on this rule, not reporting problems, read https://github.com/biomejs/biome/issues/1274#issuecomment-2407053416 for info on how to submit a detailed issue about what this rule is currently not doing, but you believe should.

**Comment by DaniGuardiola** [comment]
@pvinis it already does this, whereas the original prettier plugin respects the original whitespace. This is something that might be worth reverting though, or making optional. I haven't given it that much thought, but there might be a good reason to leave whitespace intact.

**Comment by Axelb90** [comment]
> @pvinis it already does this, whereas the original prettier plugin respects the original whitespace. This is something that might be worth reverting though, or making optional. I haven't given it that much thought, but there might be a good reason to leave whitespace intact.

Hey is not clear for me if you are planning to revert or to make it optional the white space functionality? Is it possible to know for when is this planned? 

I'm doing a refactor to a project and there are a lot of literals that cause error when running biome because the space is removed and then all became a big string 😄 . Thanks in advice I love using this tool

**Comment by pvinis** [comment]
for what it's worth, I like whitespace being time automatically for me.

**Comment by RafaelKr** [comment]
I'm currently integrating Biome into a [Symfony](https://symfony.com/) (PHP) project. Symfony uses the [Twig](https://twig.symfony.com/) template engine.

We're using Tailwind CSS and I would love to be able to apply the `useSortedClasses` rule (and especially the automatic fix) to those twig-html files.

Twig-Templates for HTML usually have the `.html.twig` file extension. And basically you could say it's a superset of HTML. So if I understand it correctly Biome first needs to stabilize HTML support, then this could be integrated, right?

But is there any way to do this currently or is it planned to be able to do that in the future?

The Symfony Demo project also uses twig with Tailwind CSS. An example twig-html file can be found at https://github.com/symfony/demo/blob/7169fb500af3be8fdcfb7badb38ac6ed88020882/templates/base.html.twig

**Comment by vdawg-git** [comment]
If I understand correctly having the classes sorted is not as important anymore with Tailwind v4, because they are using CSS layers now: https://tailwindcss.com/docs/v4-beta#native-css-cascade-layers

Sorting would be nice for consistency, but it looks like it is a lot of work to implement.

**Comment by DaniGuardiola** [comment]
CSS class order has never had an effect on precedence of styles so nothing changes with Tailwind v4.

**Comment by Mardoxx** [comment]
```
<div style="display: block; display: none;">hello</div>
<div style="display: none; display: block;">hello</div>
```

Not true.

**Comment by DaniGuardiola** [comment]
These are not CSS classes.

**Comment by m-atthart** [comment]
> ```
> 
> <div style="display: block; display: none;">hello</div>
> 
> <div style="display: none; display: block;">hello</div>
> 
> ```
> 
> 
> 
> Not true.

You're right and wrong. CSS gets applied in the order it's defined, which can be different than the order of the classes that you put on the elements.

So your example is correct, but it doesn't apply to tailwind, because in tailwind, you're not applying css to elements, you're applying classes to elements, which then get referenced in a css file. That css file has a certain order that the styles will then get applied in. So tailwind will always give the same results, regardless of the order of the classes, because the order of the css file is always the same.

The value of sorting class names is for a couple reasons. It makes the classes easier to mentally parse when you're reading someone else's code (same reason we all settled on prettier's formatting standard). And the order of the classes also happens to be the same order the classes are defined in the css file, so you can know which styles get applied in what order just by looking at the order of the classes.

If anyone doesn't fully understand, here's a good visual explainer. Watch 4:10-7:30
https://youtu.be/QBajvZaWLXs?t=250

**Comment by DaniGuardiola** [comment]
Thanks everyone for your input, let's stop this discussion though as it's off-topic to the issue at hand.

**Comment by felix-bohlin** [comment]
Are there any roadblocks in regards to the v4 release? [eslint-plugin-tailwindcss](https://github.com/francoismassart/eslint-plugin-tailwindcss/issues/325) are facing a complete rewrite since they can't rely on a `tailwind.config.js` file anymore.

**Comment by ematipico** [comment]
> Are there any roadblocks in regards to the v4 release? [eslint-plugin-tailwindcss](https://github.com/francoismassart/eslint-plugin-tailwindcss/issues/325) are facing a complete rewrite since they can't rely on a `tailwind.config.js` file anymore.

https://github.com/biomejs/biome/issues/1274#issuecomment-2407053416

**Comment by DaniGuardiola** [comment]
@felix-bohlin not particularly, it just adds a few new features we'll need to add support for, but on the other hand the new refactored v4 code is much cleaner which facilitates things, and the tailwind team is also considering implementing some APIs on their end that would facilitate the introspection of Tailwind CSS data for sorting purposes, which is nice.

**Comment by urdaydev** [comment]
Is there a way to sort tailwind classes when saving right now?

**Comment by ematipico** [comment]
@urdaydev please refer to the following page, which is reachable from the documentation of the lint rule: https://biomejs.dev/linter/#configure-the-rule-fix

**Comment by urdaydev** [comment]
@ematipico this is what my biome.json looks like but the tailwind classes are not formatted when saved, am I missing something?
```
//
"linter": {
  "enabled": true,
  "rules": {
    "recommended": true,
    "nursery": {
      "useSortedClasses": {
      	"level": "error",
      	"fix": "safe",
      	"options": {
      	  "attributes": ["className"],
      	  "functions": ["cx"]
      	}
      }
    }
  }
},
//
```

**Comment by ematipico** [comment]
@urdaydev I need to ask you to use a GitHub discussion, Discord or another issue. Your question was **already asked** here and I answered anyway, but this issue is for developments only. Please seek support using the proper channels

**Comment by DaniGuardiola** [comment]
@AlexanderOpran awesome, please reach out, ping me on the Biome discord server, I'm @ dio.la there

**Comment by carere** [comment]
Hello all :)

I don't understand how can I configure this rule to execute on save with VsCode.
I tried to add `quickFix.all = explicit` to vscode settings and it works, but with this settings, biome is really slow.
Usually I'm using `source.fixAll.biome` and biome is working as expecting, but this rule does not work...

Is there something I'm missing ? How should I configure in order to have it executed with `source.fixAll.biome` ?

**Comment by hilja** [comment]
@carere You can't, you have to press `cmd+.` and select "Sort classes".

**Comment by AndreiTS** [comment]
@carere I suggest using [Headwind](https://marketplace.visualstudio.com/items/?itemName=heybourn.headwind) for now

**Comment by carere** [comment]
Thx for the advice, I'll try Headwind 😉

**Comment by carere** [comment]
@AndreiTS I found [this extension](https://marketplace.visualstudio.com/items?itemName=dejmedus.tailwind-sorter) which take cares perfectly of the sorting 😄

**Comment by Pascalmh** [comment]
> Hello all :)
> 
> I don't understand how can I configure this rule to execute on save with VsCode.
> I tried to add `quickFix.all = explicit` to vscode settings and it works, but with this settings, biome is really slow.
> Usually I'm using `source.fixAll.biome` and biome is working as expecting, but this rule does not work...
> 
> Is there something I'm missing ? How should I configure in order to have it executed with `source.fixAll.biome` ? 

You have to mark the `useSortedClasses` as safe, then it will be applied on save: https://biomejs.dev/linter/#configure-the-rule-fix

**Comment by hilja** [comment]
> You have to mark the useSortedClasses as safe, then it will be applied on save: https://biomejs.dev/linter/#configure-the-rule-fix

I had no idea! This is great.

**Comment by sandervspl** [comment]
> > Hello all :)
> > I don't understand how can I configure this rule to execute on save with VsCode.
> > I tried to add `quickFix.all = explicit` to vscode settings and it works, but with this settings, biome is really slow.
> > Usually I'm using `source.fixAll.biome` and biome is working as expecting, but this rule does not work...
> > Is there something I'm missing ? How should I configure in order to have it executed with `source.fixAll.biome` ?
> 
> You have to mark the `useSortedClasses` as safe, then it will be applied on save: https://biomejs.dev/linter/#configure-the-rule-fix

I've never gotten this to work. Only with `quickfix.biome = explicit` does this work. What am I doing wrong?

```json
{
"useSortedClasses": {
          "level": "warn",
          "fix": "safe",
          "options": {
            "attributes": ["classList"],
            "functions": ["clsx", "cva", "tw"]
          }
        }
}
```

**Comment by Pascalmh** [comment]
@sandervspl try `"level": "error"`

like this:

```json
"useSortedClasses": {
	"level": "error",
	"fix": "safe",
	"options": {
		"functions": ["clsx", "cva", "tw"]
	}
},
```

**Comment by sandervspl** [comment]
> @sandervspl try `"level": "error"`
> 
> like this:
> 
> ```json
> "useSortedClasses": {
> 	"level": "error",
> 	"fix": "safe",
> 	"options": {
> 		"functions": ["clsx", "cva", "tw"]
> 	}
> },
> ```

Thanks, this works. Why does it only work with level: error?

**Comment by linxiano** [comment]
In my team we have the following configuration which seems to work quite decent. We haven't encountered many problems. Really hope this feature will receive more focus, to move it from the nursery group.

```
...
"linter": {
    "enabled": true,
    "rules": {
      "recommended": true,
      "a11y": {
        "noSvgWithoutTitle": "off",
        "useButtonType": "off",
        "useAltText": "off"
      },
      "nursery": {
        "useSortedClasses": {
          "fix": "safe",
          "level": "warn",
          "options": {
            "attributes": ["classList"],
            "functions": ["cn", "clsx", "cva", "tw"]
          }
        }
      }
    }
  }
...
```
Then we have format on save settings on.
```
...
  "editor.formatOnSave": true,
  "editor.defaultFormatter": "biomejs.biome",
  "editor.codeActionsOnSave": {
    "quickfix.biome": "explicit",
    "source.organizeImports.biome": "explicit"
  },
...
```

**Comment by Pedromigacz** [comment]
> In my team we have the following configuration which seems to work quite decent. We haven't encountered many problems. Really hope this feature will receive more focus, to move it from the nursery group.
> 
> ```
> ...
> "linter": {
>     "enabled": true,
>     "rules": {
>       "recommended": true,
>       "a11y": {
>         "noSvgWithoutTitle": "off",
>         "useButtonType": "off",
>         "useAltText": "off"
>       },
>       "nursery": {
>         "useSortedClasses": {
>           "fix": "safe",
>           "level": "warn",
>           "options": {
>             "attributes": ["classList"],
>             "functions": ["cn", "clsx", "cva", "tw"]
>           }
>         }
>       }
>     }
>   }
> ...
> ```
> 
> Then we have format on save settings on.
> 
> ```
> ...
>   "editor.formatOnSave": true,
>   "editor.defaultFormatter": "biomejs.biome",
>   "editor.codeActionsOnSave": {
>     "quickfix.biome": "explicit",
>     "source.organizeImports.biome": "explicit"
>   },
> ...
> ```

With this whole configuration, class sorting works on `.tsx` files but not on `.astro` files on my monorepo.
Headwind is not a workaround because it would break the rest of the project config.
Can i help in any way to make class sorting work on`.astro` files?

**Comment by Trombach** [comment]
> > In my team we have the following configuration which seems to work quite decent. We haven't encountered many problems. Really hope this feature will receive more focus, to move it from the nursery group.
> > ```
> > ...
> > "linter": {
> >     "enabled": true,
> >     "rules": {
> >       "recommended": true,
> >       "a11y": {
> >         "noSvgWithoutTitle": "off",
> >         "useButtonType": "off",
> >         "useAltText": "off"
> >       },
> >       "nursery": {
> >         "useSortedClasses": {
> >           "fix": "safe",
> >           "level": "warn",
> >           "options": {
> >             "attributes": ["classList"],
> >             "functions": ["cn", "clsx", "cva", "tw"]
> >           }
> >         }
> >       }
> >     }
> >   }
> > ...
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Then we have format on save settings on.
> > ```
> > ...
> >   "editor.formatOnSave": true,
> >   "editor.defaultFormatter": "biomejs.biome",
> >   "editor.codeActionsOnSave": {
> >     "quickfix.biome": "explicit",
> >     "source.organizeImports.biome": "explicit"
> >   },
> > ...
> > ```
> 
> With this whole configuration, class sorting works on `.tsx` files but not on `.astro` files on my monorepo. Headwind is not a workaround because it would break the rest of the project config. Can i help in any way to make class sorting work on`.astro` files?

Biome doesn't fully support astro files yet, I'll just link my earlier comment https://github.com/biomejs/biome/issues/1274#issuecomment-2329925496

As far as I'm aware there has been progress on html ish language support, but it's not there yet

**Comment by TeoChi-CF** [comment]
Great rule, but we had to disable it because of bug with literals, Biome will complain about this:
`<h1 className={`${testClass} test-${anotherTestClass}`}>`

and when sorting according to biome, it merges them together:

`<h1 className={`${testClass}test-${anotherTestClass}`}>`

**Comment by hilja** [comment]
If you do stuff like this  `<h1 className={`${testClass} test-${anotherTestClass}`}>` Tailwind doesn't know what classes to include in the final CSS.

Consider a config based approach:

```jsx
const testClasses = {
  foo: 'some-class',
  bar: 'some-other-class',
}

// Usage
<h1 className={testClasses[anotherTestClass]}>
```

Or use a simple util to combine classes:

```ts
const cn = (...classes: ClassNameValue[]) => classes.join(' ')
// Usage
cn('class1', 'class2')
```

**Comment by linxiano** [comment]
> Great rule, but we had to disable it because of bugs with literals, Biome will complain about this: `` <h1 className={`${testClass} test-${anotherTestClass}`}> ``
> 
> and when sorting according to biome, it merges them together:
> 
> `` <h1 className={`${testClass}test-${anotherTestClass}`}> ``

Also consider configure your tailwind to look for merge functions like mentioned by @hilja:
in your `.vscode/settings.json`:

```
{
...
     "tailwindCSS.classFunctions": ["cn", "cva", "cx", "clsx"],
...
}
```

**Comment by benface** [comment]
> Parasite utilities

What are "parasite utilities"?

**Comment by PlagueFPS** [comment]
One thing I would like the `useSortedClasses` option to enforce is the `lineWidth` from the formatted. Currently a long classname is not formatted into a new or multiple new lines, instead since the whitespace is collapsed into a single space it is just one super long line.

I'd like there to be an option to use the default config or enforce the `lineWidth` format config. Personally I do not like the one single long line, I'd like to break it into a new line or multiple if needed.

**Comment by FunctionDJ** [comment]
Would love to use Biome but automatic line breaks are a must-have for me. I'll keep using Prettier until then.
I subscribed to the issue though :eyes:

**Comment by leslie555** [comment]
my settings.json
```
{
  "editor.formatOnSave": true,
  "editor.defaultFormatter": "biomejs.biome",
  "editor.codeActionsOnSave": {
    "source.fixAll.biome": "explicit",
    "source.organizeImports.biome": "explicit"
  }
}
```

my biome.json
```
 "nursery": {
        "useSortedClasses": {
          "level": "error",
          "fix": "safe",
          "options": {
            "attributes": ["className"],
            "functions": ["cn", "cva", "tw", "clsx"]
          }
        }
      }
```

My problem is when I save my file, the code will be formatted like the code below:
`<ChevronDownIcon className='opacity-50 size-4' />`
But it will report an error since I configured **useSortedClasses**.  _size-4_ should be placed forward of _opacity_50_ . How to solve it?

**Comment by akira1ce** [comment]
> my settings.json
> 
> ```
> {
>   "editor.formatOnSave": true,
>   "editor.defaultFormatter": "biomejs.biome",
>   "editor.codeActionsOnSave": {
>     "source.fixAll.biome": "explicit",
>     "source.organizeImports.biome": "explicit"
>   }
> }
> ```
> 
> my biome.json
> 
> ```
>  "nursery": {
>         "useSortedClasses": {
>           "level": "error",
>           "fix": "safe",
>           "options": {
>             "attributes": ["className"],
>             "functions": ["cn", "cva", "tw", "clsx"]
>           }
>         }
>       }
> ```
> 
> My problem is when I save my file, the code will be formatted like the code below: `<ChevronDownIcon className='opacity-50 size-4' />` But it will report an error since I configured **useSortedClasses**. _size-4_ should be placed forward of _opacity_50_ . How to solve it?

are u using biome 2.1.2 ?

2.1.1 has a un-safe fix, but i can not found it in  2.1.2 .

may be u can try 2.1.1, then it will be fixed when save.

**Comment by leslie555** [comment]
> are u using biome 2.1.2 ?
> 
> 2.1.1 has a un-safe fix, but i can not found it in 2.1.2 .
> 
> may be u can try 2.1.1, then it will be fixed when save.

Thanks, but the issue still exists. I think it's a code format issue. I didn't find out the proper config yet.

**Comment by akira1ce** [comment]
> > are u using biome 2.1.2 ?
> > 2.1.1 has a un-safe fix, but i can not found it in 2.1.2 .
> > may be u can try 2.1.1, then it will be fixed when save.
> 
> Thanks, but the issue still exists. I think it's a code format issue. I didn't find out the proper config yet.谢谢，但问题仍然存在。我认为这是代码格式问题，我还没有找到正确的配置。

I got.

        "useSortedClasses": {
          "level": "error",
          "fix": "safe",
          "options": {
            "attributes": ["classList", "className"],
            "functions": ["clsx", "cva", "tw", "tw.*"]
          }
        }

#6982 

your config is same,haha.

but this can solve my error

**Comment by dvzrd** [comment]
> > Parasite utilities
> 
> What are "parasite utilities"?

I'm not sure, but could be related to a css formatting issue when using tailwindcss v4 `@utility` to create functional utilities. I had to define my utilities inside separate files so I can ignore them like this:

```
 "files": {
    "includes": [
      "**",
      "!**/utilities.css",
    ]
  },
```

There could be syntax related to calculations that may be allowed in tailwind but not recognized by base css syntax.

**Comment by dyc3** [comment]
@dvzrd That's correct, Biome's CSS parser does not support all of Tailwind's things right now. See also: #3195

**Comment by DaniGuardiola** [comment]
Parasite utilities are classes that don't have a direct purpose and don't apply any styles themselves. They only exist to provide value to other classes. An example of this is "group" (a "parasite" utility), which by itself does not do anything at all, but is necessary for "group-hover:" and similar variants.

**Comment by elia** [comment]
Not sure if this is where I should report this, but I had an instance in which using single-quotes in TSX generated an invalid `className` attribute because it didn't convert `'size-'` inside `[&_svg:not([class*='size-'])]:size-4` to double quotes.

**Comment by dyc3** [comment]
@elia Sounds like a bug. Feel free to open a new issue.

**Comment by JKd3vLD** [comment]
I don't know if this is a bug and if it was reported before, but I found something while using the
```
"useSortedClasses": {
          "fix": "safe"
}
```
When I use it with the 
```
"javascript": {
    "formatter": {
        "jsxQuoteStyle": "double",
        "quoteStyle": "single"
    }
}
```
The jsx or tsx component className attribute defaults to using single quotes when sorting the classes instead of using the same as jsxQuoteStyle.

**Comment by ricohumme** [comment]
I just ran into an issue where whitespace is being collapsed where I feel it is not meant to.

We have not added rule `useSortedClasses` to our biome configuration, so it is running at the default.
The following case is being collapsed:

```jsx
<ul className="steps m-auto">
  {[...new Array(totalSteps)].map((_, i) => (
    <li
      className={`step${i + 1 <= currentStepIndex ? ' step-primary' : ''}`}
      key={i}
    />
  ))}
</ul>
```

When we format the code, `' step-primary'` becomes `'step-primary'`, causing the output for this class to become `stepstep-primary`, which breaks our UI.
We could ofcourse, which we will do for the time being, add a space behind the first step:
```jsx
`step ${i + 1 <= currentStepIndex ? 'step-primary' : ''}`
```
But this in turn would have the effect that for every step there is a trailing space behind step in the generated html, which is just plain-old ugly if you ask me.

Hopefully, this can be addressed. I do support the feature of sorting classes, just not the whitespace trimming when applied with logic

**Comment by dyc3** [comment]
@ricohumme IMO, your example is an antipattern. You should likely be using [tailwind-merge](https://www.npmjs.com/package/tailwind-merge) for conditional styling, or something similar. **Tailwind doesn't like it when you dynamically build strings for it's classes**, whether its inside a template or doing something like `class"text-red-" + "500"`. It wasn't made for that. 

Tailwind overwhelmingly prefers that you explicitly match the class names that they provide in their docs. In other words, they greatly prefer that anything that is in a `class` attribute is a string literal, and not anything that is dynamically built.

**Comment by boar-is** [comment]
@dyc3, there's nothing wrong with @ricohumme's example from Tailwind perspective. He doesn't build strings dynamically. It's basically:
```js
i + 1 <= currentStepIndex ? 'step step-primary' : 'step'
```

@ricohumme, as a temporary workaround I think this might help:
```js
const whitespace = ' '

className={`step${i + 1 <= currentStepIndex ? `${whitespace}step-primary` : ''}`}
```

**Comment by hilja** [comment]
Or:
```
className={`step${i + 1 <= currentStepIndex ? `${' '}step-primary` : ''}`}
```

**Comment by DaniGuardiola** [comment]
@ricohumme's case is a bit on the edge. i agree it's not ideal.

maybe a good middle-ground solution is to just ignore any inserted expressions within template strings. can anyone think of an example where having a boundary there for the sorter could be problematic?

**Comment by eliotgevers** [comment]
Is Tailwind V4 supported? Because I saw `tailwind.config.js` mentioned, but V4 doesn't use that.

**Comment by Azarattum** [comment]
Is it not supposed to work in plain HTML files (and HTML embedded languages yet)?

```
$ echo '<div class="px-2 foo p-4 bar"></div>' > tmp.html
$ bun biome lint tmp.html                               
Checked 1 file in 696µs. No fixes applied.
```

Is there any way to use this rule outside of React/JSX at all?

**Comment by dyc3** [comment]
@Azarattum not yet. it has to be ported from `biome_js_analyze` to `biome_html_analyze`

**Comment by balazshevesi** [comment]
Because this thread is huge, here's a **quick guide and recap to anyone setting class-sorting for tailwind up in their project and just want to get it working ASAP:**

## **Recap:**

- The port of `prettier-plugin-tailwindcss` to biome is still work in progress
  - Therefore it's currently part of the `nursery` config group [docs](https://biomejs.dev/linter/#nursery)
- Currently only linting (*not formatting*) of class ordering is supported
  - *But* you can run a "code action" which *will* format the order of the classes
  - `source.fixAll.biome` is the name of the code action which will perform "Safe fixes" ("Safe fixes are guaranteed to not change the semantic of your code. They can be applied without explicit review. - [docs](https://biomejs.dev/linter/#safe-fixes)").
    - <del>Class sorting is part of the safe fixes</del>. By default class sorting *is not* part of the safe fixes, but can set to be considered a safe fix by `"fix": "safe"` in the config.

## **Steps to get it working:** 

### 1. Enable `useSortedClasses` in your `biome.json`:

Here's my config for linting (stolen shamelessly from @linxiano)
```json
"linter": {
  "enabled": true,
  "rules": {
    "recommended": true,
    "a11y": {
      "noSvgWithoutTitle": "off",
      "useButtonType": "off",
      "useAltText": "off"
    },
    "nursery": {
      "useSortedClasses": {
        "fix": "safe",
        "level": "warn",
        "options": {
          "attributes": ["classList"],
          "functions": ["cn", "clsx", "cva", "tw"]
        }
      }
    }
  }
},
```

### 2. Figure out a way to run `source.fixAll.biome` on format (this will depend on your editor of choice)

#### Zed

For me, I use [Zed](https://zed.dev/), I use this is my `settings.json`
```json
"languages": {
  "JavaScript": {
    "formatter": { "language_server": { "name": "biome" } },
    "code_actions_on_format": {
      "quickfix": true,
      "source.fixAll.biome": true
    }
  },
  "TypeScript": {
    "formatter": { "language_server": { "name": "biome" } },
    "code_actions_on_format": {
      "quickfix": true,
      "source.fixAll.biome": true
    }
  },
  "TSX": {
    "formatter": { "language_server": { "name": "biome" } },
    "code_actions_on_format": {
      "quickfix": true,
      "source.fixAll.biome": true
    }
  },
  "JSON": { "formatter": { "language_server": { "name": "biome" } } },
  "JSONC": { "formatter": { "language_server": { "name": "biome" } } },
  "CSS": { "formatter": { "language_server": { "name": "biome" } } }
},
```
(Edit: setting `"quickfix": true` will preform the quick fixes on the code-editor level, you can set it to `false` if you only want class-formatting from biome)

#### Visual Studio Code

On VSC (also in `settings.json`) it should be the following:
```json
"editor.formatOnSave": true,
"editor.codeActionsOnSave": {
  "source.fixAll.biome": "explicit",
  "source.organizeImports.biome": "explicit"
}
```

I hope this helps!

**Comment by PaulRBerg** [comment]
Hey @balazshevesi thanks for the recap and the copy-pasteable config.

Just a correction on what you said:

- By default, Biome considers the fixes for `useSortedClasses` as **unsafe** (although this might change in the future)
- However, as users, we can override the safe/unsafe status of any linter rule
- In your config, you marked the rule as **safe** when you wrote this:

```json
"fix": "safe"
```

**Comment by balazshevesi** [comment]
@PaulRBerg You're correct. I edited it and clarified it

**Comment by eaglah** [comment]
Would you plan to add support for Svelte and HTML?

**Comment by jakeleventhal** [comment]
`text-x` and `leading-y` classes can be combined to `text-x/y` (see https://github.com/tailwindlabs/tailwindcss/pull/19396)

This should be part of the tailwind autoformatting in Biome

**Comment by jakeleventhal** [comment]
I also believe all the rules captured in tailwinds `suggestCanonicalClasses` should be captured by this lint rule

**Comment by hqw567** [comment]
## Feature request: Regex/wildcard pattern support for `attributes` option

### Problem

The `attributes` option currently only supports exact string matching. This makes it impractical for codebases that follow a consistent naming convention for class-related props (e.g. `boxClassName`, `wrapClassName`, `inputClassName`, `labelClassName`).

Right now every custom attribute must be listed individually:

```json
"attributes": ["classList", "boxClassName", "wrapClassName", "inputClassName", "labelClassName"]
```

This is tedious to maintain — it's easy to forget to add a new attribute when adding new components.

### Real-world use case

When building a component library, it's common to expose multiple class injection points per component:

```tsx
<SkewInput
  className="..."
  boxClassName="bg-red h-13 w-full"
  labelClassName="text-sm font-medium"
  wrapClassName="flex flex-col gap-2"
/>
```

Without pattern support, `useSortedClasses` silently skips all custom attributes unless each one is explicitly listed — easy to miss and hard to scale.

### Proposed solution

Support regex patterns in the `attributes` option, similar to how `prettier-plugin-tailwindcss` handles its `tailwindAttributes` option (enclosed in `/` slashes):

```json
"attributes": ["classList", "/.*[Cc]lass.*/"]
```

Alternatively, a simpler glob-style wildcard (consistent with the existing `tw.*` syntax in `functions`) would also work:

```json
"attributes": ["classList", "*ClassName"]
```

### Prior art

`prettier-plugin-tailwindcss` supports regex patterns in its `tailwindAttributes` option:

```json
// .prettierrc
{
  "tailwindAttributes": ["myClassList", "/data-.*/"]
}
```

This matches any attribute starting with `data-`, making it easy to handle entire naming conventions without listing each attribute explicitly.

It would be great to have equivalent support in Biome!

**Comment by Lihu0** [comment]
Are there plans of adding HTML support to this? I want to use this for Astro but I'm not doing it since I use TailwindCSS

**Comment by gameroman** [comment]
> Are there plans of adding HTML support to this? I want to use this for Astro but I'm not doing it since I use TailwindCSS

See #11102

**Comment by johncarmack1984** [comment]
Status update from my end, since I closed #11396 / #11399 / #11401 last week and didn't say why here.

Where the sorting half of this stands: the v4 engine is done and matches getClassOrder on the probe set as of #11357 (that one closed the last gap, the legacy leading !). Variant parsing, arbitrary values, modifiers, container queries are all in main already.

What is NOT done is custom classes: anything from `@utility` / `@custom-variant` / `@theme` or a plugin classifies as unknown and gets left at the front, which is where the prettier plugin and this rule diverge on any real project. I don't think the v4 engine should be wired in until that's closed, and the stylesheet-reading branch that closes it is built locally. Also still open: parser recovery on a malformed class (right now one bad class swallows the rest of the line). That one's @dyc3 's per feedback #11357, I've got the standalone version at 56ad02f78e if it's useful, otherwise I'll stay clear of it.

I pulled the PRs over schedule limitations I had last week that are now resolved; I'm good to see this issue through this week & next, provided @dyc3 and @ematipico are satisfied with the coordination and implementation.

If there's still appetite for this, I can put the branches back up in whatever shape is easiest to review: engine move to `biome_tailwind_logic` (#11396 as it was), the `tailwind.stylesheet` config surface (#11399), and the wiring last. @dyc3 your call on order and whether the config half should wait on your `tailwind-configuration` branch + how I can help if you're bandwidth-limited, and @ematipico I'd defer to you on whether #11401 is an acceptable way to write fixes into Astro templates.
